dtm0/fop.c

```diff
@@ -0,0 +1,392 @@
+	[M0_FOPH_DTM0_LOGGING] = {
+		.sd_name      = "logging",
+		.sd_allowed   = M0_BITS(M0_FOPH_SUCCESS,
+					M0_FOPH_FAILURE)
+	},
+};
+
+struct m0_sm_trans_descr dtm0_phases_trans[] = {
+	[ARRAY_SIZE(m0_generic_phases_trans)] =
+	{"dtm0_1-fail", M0_FOPH_TYPE_SPECIFIC, M0_FOPH_FAILURE},
+	{"dtm0_1-success", M0_FOPH_TYPE_SPECIFIC, M0_FOPH_SUCCESS},
+};
+
+static struct m0_sm_conf dtm0_conf = {
+	.scf_name      = "dtm0-fom",
+	.scf_nr_states = ARRAY_SIZE(dtm0_phases),
+	.scf_state     = dtm0_phases,
+	.scf_trans_nr  = ARRAY_SIZE(dtm0_phases_trans),
+	.scf_trans     = dtm0_phases_trans,
+};
+
+M0_INTERNAL int m0_dtm0_fop_init(void)
+{
+	static int init_once = 0;
+	if (init_once++ > 0)
+		return 0;
+
+	m0_sm_conf_extend(m0_generic_conf.scf_state, dtm0_phases,
+			  m0_generic_conf.scf_nr_states);
+	m0_sm_conf_trans_extend(&m0_generic_conf, &dtm0_conf);
+	m0_sm_conf_init(&dtm0_conf);
+
+
+	m0_xc_dtm0_fop_init();
+	M0_FOP_TYPE_INIT(&dtm0_req_fop_fopt,
+			 .name      = "DTM0 request",
+			 .opcode    = M0_DTM0_REQ_OPCODE,
+			 .xt        = dtm0_req_fop_xc,
+			 .rpc_flags = M0_RPC_MUTABO_REQ,
+			 .fom_ops   = &dtm0_req_fom_type_ops,
+			 .sm        = &dtm0_conf,
+			 .svc_type  = &dtm0_service_type);
+	M0_FOP_TYPE_INIT(&dtm0_rep_fop_fopt,
+			 .name      = "DTM0 reply",
+			 .opcode    = M0_DTM0_REP_OPCODE,
+			 .xt        = dtm0_rep_fop_xc,
+			 .rpc_flags = M0_RPC_ITEM_TYPE_REPLY,
+			 .fom_ops   = &dtm0_req_fom_type_ops);
+	return m0_fop_type_addb2_instrument(&dtm0_req_fop_fopt);
+}
+
+/*
+  Allocates a fop.
+ */
+
+int m0_dtm0_fop_create(struct m0_rpc_session	 *session,
+		       enum m0_dtm0s_msg	  opmsg,
+		       struct m0_dtm0_tx_desc	 *txr,
+		       struct m0_fop		**out)
+{
+	struct dtm0_req_fop  *op;
+	struct m0_fop	     *fop;
+	int                rc = 0;
+
+	*out = NULL;
+
+	fop = m0_fop_alloc_at(session, &dtm0_req_fop_fopt);
+	if (fop == NULL)
+		return M0_ERR(-ENOMEM);
+	op = m0_fop_data(fop);
+	op->dtr_msg = opmsg;
+	op->dtr_txr = *txr;
+	*out = fop;
+	return rc;
+}
+M0_EXPORTED(m0_dtm0_fop_create);
+
+/*
+  Allocates a fom.
+ */
+static int dtm0_fom_create(struct m0_fop *fop,
+			   struct m0_fom **out, struct m0_reqh *reqh)
+{
+	struct dtm0_fom          *fom;
+	struct m0_fop            *repfop = NULL;
+	struct dtm0_rep_fop      *reply;
+	struct m0_dtm0_pmsg_ast  *pma = NULL;
+
+	M0_ENTRY("reqh=%p", reqh);
+
+	M0_ALLOC_PTR(fom);
+	M0_ALLOC_PTR(pma);
+	repfop = m0_fop_reply_alloc(fop, &dtm0_rep_fop_fopt);
+
+	if (fom != NULL && repfop != NULL && pma != NULL) {
+		*out = &fom->dtf_fom;
+		/** TODO: calculate credits for the operation.
+		 */
+
+		/* see ::m0_dtm0_dtx_post_persistent for the details */
+		fop->f_opaque = pma;
+		reply = m0_fop_data(repfop);
+		reply->dr_txr = (struct m0_dtm0_tx_desc) {};
+		reply->dr_rc = 0;
+		m0_fom_init(&fom->dtf_fom, &fop->f_type->ft_fom_type,
+			    &dtm0_req_fom_ops, fop, repfop, reqh);
+		return M0_RC_INFO(0, "fom=%p", &fom->dtf_fom);
+	} else {
+		m0_free(pma);
```

Comment:
Freeing of NULL is possible here. I don't remember what exactly m0_free() does in this case and better to check it out or rework this place a bit.

Reply:
m0_free() (like free() from libc's starting from quite old versions) checks for NULL, so this is safe.